Port OMPL planning plugin

henningkayser opened the pull request:
I performed all migration steps (except for dynamic_reconfigure) and am now testing/fixing this with MoveItCpp #163. I expect that getting this to run is mostly an issue of loading the correct parameters.

mamoll commented on moveit_planners/ompl/CMakeLists.txt:
```diff
@@ -42,5 +38,12 @@ include_directories(ompl_interface/include
+
+ament_export_include_directories(include)
+ament_export_libraries(moveit_ompl_interface)
+ament_export_dependencies(moveit_core OMPL)
```
Shouldn't that be `ompl`? Where is this defined?

henningkayser replied:
You're probably right. However, I'm not sure how ament handles this since the include/library targets have the prefix `OMPL` which is why also `ament_target_dependencies()` only works with `OMPL`.

henningkayser replied:
Also, `OMPL` was used for catkin_package `DEPENDS` as well.

henningkayser commented:
I would like to avoid growing this PR any bigger. Currently, all migration steps are applied (logging, message types, node handling). CI and tests are passing. I'm planning to add parameter lookup in a follow-up PR since that is related to changes in the planning pipeline and interface headers.

mlautman commented on moveit_planners/ompl/ompl_interface/src/ompl_interface.cpp:
```diff
@@ -232,27 +224,11 @@ void ompl_interface::OMPLInterface::loadPlannerConfigurations()
-    XmlRpc::XmlRpcValue config_names;
-    if (nh_.getParam(group_name + "/planner_configs", config_names))
+    std::vector<std::string> config_names;
+    if (node_->get_parameter(group_name + ".planner_configs", config_names))
```
Are parameters separated by "/" or "." in ros2? This PR seems to use both

henningkayser replied:
It's ".". As pointed out in the description this PR only contains straight migration steps, I'm providing a thorough parameter refactoring in a follow-up PR. I can change these for cleanliness, though.

mlautman commented on moveit_planners/ompl/ompl_interface/src/ompl_planner_manager.cpp:
```diff
@@ -79,16 +84,16 @@ class OMPLPlannerManager : public planning_interface::PlannerManager
           case ompl::msg::LOG_DEV2:
           case ompl::msg::LOG_DEV1:
           case ompl::msg::LOG_DEBUG:
-            OMPL_ROS_LOG(::ros::console::levels::Debug);
```
@mamoll Do you know why this package uses the different Logging macros?

mlautman replied:
I'm all for standardizing to the RCLCPP ones as they filter better

henningkayser replied:
I think the usual ROS logging macros wouldn't allow you to change the logging name (i.e. filename, line) because it's stored in a static variable. I'm not sure if the RCLCPP approach I added prints nice debug messages, but I'll test this when running the MoveItCpp integration

mamoll replied:
@mlautman OMPL is trying to be ROS-agnostic. It has its logging infrastructure, but can be easily hooked into the ROS logging infrastructure. (Not a coincidence, BTW, since Ioan Sucan wrote the original version of logging for both, I believe.)

mlautman commented on moveit_planners/ompl/ompl_interface/src/ompl_planner_manager.cpp:
```diff
@@ -102,16 +107,15 @@ class OMPLPlannerManager : public planning_interface::PlannerManager
   }
 
-  bool initialize(const robot_model::RobotModelConstPtr& model, const std::string& ns) override
+  bool initialize(const robot_model::RobotModelConstPtr& model, const rclcpp::Node::SharedPtr& node) override
```
could be teplated

mlautman commented on moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp:
```diff
@@ -289,25 +290,26 @@ const ompl_interface::ModelBasedStateSpaceFactoryPtr& ompl_interface::PlanningCo
 ompl_interface::ModelBasedPlanningContextPtr ompl_interface::PlanningContextManager::getPlanningContext(
     const planning_scene::PlanningSceneConstPtr& planning_scene, const moveit_msgs::msg::MotionPlanRequest& req,
-    moveit_msgs::msg::MoveItErrorCodes& error_code, const ros::NodeHandle& nh, bool use_constraints_approximation) const
+    moveit_msgs::msg::MoveItErrorCodes& error_code, const rclcpp::Node::SharedPtr& node,
```
could be templated

henningkayser commented on moveit_core/kinematic_constraints/include/moveit/kinematic_constraints/utils.h:
```diff
@@ -196,13 +196,14 @@ moveit_msgs::msg::Constraints constructGoalConstraints(const std::string& link_n
  */
-// TODO rework this function
-// bool constructConstraints(XmlRpc::XmlRpcValue& params, moveit_msgs::msg::Constraints& constraints);
+bool constructConstraints(const rclcpp::Node::SharedPtr& node, const std::string& constraints_param,
```
Next PR... parameter lookup won't be part of this, since this requires refactoring in planning pipeline, planning interface headers etc

henningkayser replied:
I added more documentation how the yaml file should look like

henningkayser commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -281,263 +283,248 @@ moveit_msgs::msg::Constraints constructGoalConstraints(const std::string& link_n
 }
 
-// TODO ROS 2: Rework these functions
+static bool constructPoseStamped(const rclcpp::Node::SharedPtr& node, const std::string& pose_param,
```
Just as the original function, this is just a local (static) helper function for loading constraints.

henningkayser commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -281,263 +283,248 @@ moveit_msgs::msg::Constraints constructGoalConstraints(const std::string& link_n
+  return true;
+}
+
+bool constructConstraints(const rclcpp::Node::SharedPtr& node, const std::string& constraints_param,
```
No, this non-static function is declared and documented in the header file. All static functions are just local helper functions which are not publicly accessible.

henningkayser commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/model_based_planning_context.h:
```diff
@@ -317,13 +317,13 @@ class ModelBasedPlanningContext : public planning_interface::PlanningContext
   /** @brief Look up param server 'constraint_approximations' and use its value as the path to load constraint
    * approximations to */
-  bool loadConstraintApproximations(const ros::NodeHandle& nh);
+  bool loadConstraintApproximations(const rclcpp::Node::SharedPtr& node);
```
Will consider it when refactoring parameter lookup

henningkayser commented on moveit_planners/ompl/ompl_interface/scripts/generate_state_database.cpp:
```diff
@@ -130,23 +151,22 @@ void computeDB(const planning_scene::PlanningScenePtr& scene, struct GenerateSta
 
-  ROS_INFO_STREAM_NAMED(LOGNAME, "Generating Joint Space Constraint Approximation Database for constraint:\n"
-                                     << params.constraints);
+  RCLCPP_INFO_STREAM(LOGGER, "Generating Joint Space Constraint Approximation Database for constraint:\n"
```
You mean the stream macro? It's new in eloquent

henningkayser commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -281,263 +283,248 @@ moveit_msgs::msg::Constraints constructGoalConstraints(const std::string& link_n
+static bool constructPoseStamped(const rclcpp::Node::SharedPtr& node, const std::string& pose_param,
+                                 geometry_msgs::msg::PoseStamped& pose)
+{
+  if (!node->get_parameter(pose_param + "frame_id", pose.header.frame_id))
```
It should be `+ ".frame_id"`, good catch

henningkayser commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -281,263 +283,248 @@ moveit_msgs::msg::Constraints constructGoalConstraints(const std::string& link_n
+  return true;
+}
+
+static bool constructConstraint(const rclcpp::Node::SharedPtr& node, const std::string& constraint_param,
```
No, it's a local helper function and not publicly used at all. I added better documentation, though

henningkayser commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -281,263 +283,248 @@ moveit_msgs::msg::Constraints constructGoalConstraints(const std::string& link_n
+static bool constructConstraint(const rclcpp::Node::SharedPtr& node, const std::string& constraint_param,
+                                moveit_msgs::msg::JointConstraint& constraint)
+{
+  node->get_parameter(constraint_param + "weight", constraint.weight);
```
Originally, I had `constraint_param` include the `"."` but I realized this is confusing. I changed it so that the `.` is always included in the strings.

henningkayser commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -281,263 +283,248 @@ moveit_msgs::msg::Constraints constructGoalConstraints(const std::string& link_n
-// }  // namespace kinematic_constraints
+  std::vector<double> orientation;
+  if (!node->get_parameter(pose_param + "orientation", orientation) || orientation.size() != 3)
+    return false;
```
There is logging that warns about wrong format. As this is a straight port I don't think it makes sense to add too much extra logging, especially since we should switch to declaring parameters which would allow better and more verbose checking at runtime.

henningkayser commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -281,263 +283,248 @@ moveit_msgs::msg::Constraints constructGoalConstraints(const std::string& link_n
+  return true;
+}
+
+static bool constructConstraint(const rclcpp::Node::SharedPtr& node, const std::string& constraint_param,
```
It's not used anywhere except for here so I don't think it should be made public.

henningkayser commented:
@mlautman I think I addressed your points, thanks for pointing out the wrong parameter names and lacking documentation. I think changing the node interfaces is out of scope of this PR (this is a straight migration) and should be considered when actually running this code and loading yaml files.

davetcoleman commented on moveit2.repos:
```diff
@@ -47,3 +47,10 @@ repositories:
     version: ros2
+
+# Temporarily install OMPL from melodic
+# TODO(henningkayser): remove once OMPL is released for ROS2
```
Love this note. Does @mamoll know about this todo? He releases OMPL

henningkayser replied:
He will, in the release announcement ;)

davetcoleman commented on moveit_planners/ompl/ompl_interface/src/ompl_planner_manager.cpp:
```diff
@@ -260,53 +264,54 @@ class OMPLPlannerManager : public planning_interface::PlannerManager
-  ros::Publisher pub_valid_traj_;
+  // rclcpp::Publisher pub_markers_;
+  // rclcpp::Publisher pub_valid_states_;
+  // rclcpp::Publisher pub_valid_traj_;
```
This stuff has been cruft for a long time and should just be removed.
